README.md
# Aramask 🦅🦊

Aragon account plugin for Metamask.

It allows DAO users to add a special contract account to their Metamask wallet and use it to manage [Aragon Agent](https://help.aragon.org/article/37-agent). It is based on a new [Metamask Plugin System](https://github.com/MetaMask/metamask-snaps-beta/wiki/Motivation).

## Usage

Current limitations:

- **Needs [experimental version](https://github.com/MetaMask/metamask-snaps-beta) of Metamask.**
- Which works only in Chrome/Chromium and Brave *(Note that in Brave you may have to disable default Crypto Wallet in order to use Metamask)*.
- Needs backend service to construct Agent transactions.
- Metamask will [display agent's contract address](https://github.com/xuhcc/aramask/issues/2) in its account list instead of DAO address.
- Every transaction requires two confirmations from the user.
- Successful transactions will be [marked as failed](https://github.com/xuhcc/aramask/issues/3) in transaction list of Aragon account.
- Agent account will not be preserved after browser restart.

### Metamask installation

Clone the [metamask-snaps-beta](https://github.com/MetaMask/metamask-snaps-beta) repo.

Use the `develop` branch.

Install required packages and build Metamask-Beta:

```
yarn install
yarn dist
```

Load unpacked extension from `dist/chrome/`.

### Demo app and plugin installation

Clone this repo.

Install required packages and build the plugin:

```
npm install
npm run build
```

Start server:

```
npm start
```

Navigate to http://localhost:8081/, follow the instructions.


CHANGELOG.md
# Changelog

## 0.2.0

- DAO account can be added by ENS name.

## 0.1.0

- Initial release.


test.js
#!/usr/bin/env node

const agent = require('./agent')

const args = process.argv

const daoAddress = args[2]
const actorAddress = args[3]
const txParams = [args[4], args[5], args[6]]

agent.calculatePath(daoAddress, actorAddress, txParams).then(() => {
    process.exit()
})


app.js
const snapId = new URL('package.json', window.location.href).toString()

const installButton = document.querySelector('button.install')
const addAccountButton = document.querySelector('button.add-account')
const connectButton = document.querySelector('button.connect')
const setActorButton = document.querySelector('button.set-actor')
const sendButton = document.querySelector('button.send')

installButton.addEventListener('click', install)
addAccountButton.addEventListener('click', addAccount)
connectButton.addEventListener('click', connect)
setActorButton.addEventListener('click', setActor)
sendButton.addEventListener('click', sendTransaction)

async function install() {
    console.log('installing...')
    await ethereum.send({
        method: 'wallet_enable',
        params: [{
            'wallet_plugin': {
                [snapId]: {},
            },
            'eth_accounts': {},
        }],
    })
    console.log('installed.')
}

async function connect() {
    console.log('connecting to account...')
    await ethereum.send({
        method: 'wallet_enable',
        params: [{
            'eth_accounts': {},
        }],
    })
    console.log('connected.')
}

async function addAccount() {
    const daoName = document.querySelector('input.dao-name').value.toLowerCase()
    console.log('looking for agent in dao: ', daoName)
    const statusBox = document.querySelector('.add-account-status')
    statusBox.textContent = 'please wait...'
    statusBox.style.color = 'green'
    let account
    try {
        account = await ethereum.send({
            method: 'wallet_invokePlugin',
            params: [snapId, {
                method: 'addAccount',
                params: [
                    daoName,
                ],
            }],
        })
    } catch (error) {
        statusBox.textContent = error.message
        statusBox.style.color = 'red'
        console.error(error)
        return
    }
    statusBox.textContent = `account added: ${account}`
    console.log('account added: ', account)
}

async function setActor() {
    const accounts = await ethereum.send({
        method: 'eth_accounts',
    })
    if (accounts.length === 0) {
        alert('no accounts available.')
        return
    }
    const account = accounts[0]
    const actor = document.querySelector('input.actor-address').value.toLowerCase()
    console.log('setting actor: ', actor)
    const statusBox = document.querySelector('.set-actor-status')
    try {
        await ethereum.send({
            method: 'wallet_invokePlugin',
            params: [snapId, {
                method: 'setActor',
                params: [
                    account,
                    actor,
                ],
            }],
        })
    } catch (error) {
        statusBox.textContent = error.message
        statusBox.style.color = 'red'
        console.error(error)
        return
    }
    statusBox.textContent = `done.`
    statusBox.style.color = 'green'
    console.log('actor set.')
}

async function sendTransaction() {
    const accounts = await ethereum.send({
        method: 'eth_accounts',
    })
    if (accounts.length === 0) {
        alert('no accounts available.')
        return
    }
    const account = accounts[0]
    console.log('sending transaction...')
    const value = 0
    const to = '0x0000000000000000000000000000000000000000'
    const response = await ethereum.send({
        method: 'eth_sendTransaction',
        params: [{
            from: account,
            to: to,
            value: value.toString(16),
        }],
    })
    console.log('sent: ', response)
}


agent.js
const { default: Aragon, ensResolve } = require('@aragon/wrapper')
const Web3 = require('web3')
const namehash = require('eth-ens-namehash')

const ARAGON_ETH_PROVIDERS = {
    1: 'wss://mainnet.eth.aragon.network/ws',
    4: 'wss://rinkeby.eth.aragon.network/ws',
}
const ARAGON_IPFS_GATEWAY = 'https://ipfs.eth.aragon.network/ipfs'
const ENS_REGISTRIES = {
    // https://github.com/aragon/aragon/blob/master/arapp.json
    1: '0x00000000000C2E074eC69A0dFb2997BA6C7d2e1e',
    4: '0x98df287b6c145399aaa709692c8d308357bc085d',
}
const AGENT_APP_ID = namehash.hash('agent.aragonpm.eth')

async function getDao(chainId, daoName) {
    const ethProvider = new Web3.providers.WebsocketProvider(
        process.env.ETHEREUM_URL || ARAGON_ETH_PROVIDERS[chainId],
    )
    const ipfsGatewayUrl = process.env.IPFS_GATEWAY_URL || ARAGON_IPFS_GATEWAY
    const ensRegistryAddress = ENS_REGISTRIES[chainId]

    let daoAddress
    if (daoName.startsWith('0x') && daoName.length === 42) {
        daoAddress = daoName
    } else {
        if (!daoName.endsWith('.eth')) {
            daoName += '.aragonid.eth'
        }
        daoAddress = await ensResolve(daoName, {
            provider: ethProvider,
            registryAddress: ensRegistryAddress,
        })
        console.log(`Resolved ${daoName} to ${daoAddress}`)
    }

    const dao = new Aragon(daoAddress, {
        provider: ethProvider,
        apm: {
            ipfs: {
                gateway: ipfsGatewayUrl,
            },
            ensRegistryAddress,
        },
    })
    await dao.init()
    console.log(`Found DAO at ${daoAddress}`)
    return dao
}

function getAgentAddress(dao) {
    return new Promise((resolve, reject) => {
        const subscription = dao.apps.subscribe((apps) => {
            subscription.unsubscribe()
            for (let idx = 0; idx < apps.length; idx++) {
                let app = apps[idx]
                if (app.appId === AGENT_APP_ID) {
                    const agentAddress = app.proxyAddress
                    console.log(`Agent address is ${agentAddress}`)
                    resolve(agentAddress)
                    return
                }
            }
            reject('agent not found')
        })
    })
}

async function findAgent(chainId, daoName) {
    const dao = await getDao(chainId, daoName)
    const agentAddress = await getAgentAddress(dao)
    return {
        daoAddress: dao.kernelProxy.address,
        agentAddress,
    }
}

async function calculatePath(chainId, daoAddress, actorAddress, txParams) {
    /*
     * txParams: [to, value, data]
     */
    const dao = await getDao(chainId, daoAddress)
    const agentAddress = await getAgentAddress(dao)
    // Forward transaction to an agent
    const result = await dao.calculateTransactionPath(actorAddress, agentAddress, 'execute', txParams)
    const tx = result[0]
    console.log('Forwarded TX: ', tx)
    return tx
}

module.exports.findAgent = findAgent
module.exports.calculatePath = calculatePath


index.js
const BACKEND_URL = 'http://localhost:8084'

const accounts = {}

wallet.registerRpcMessageHandler(async (originString, requestObject) => {
    switch (requestObject.method) {
        case 'addAccount':
            return await addAccount(requestObject.params[0])
        case 'setActor':
            await setActor(...requestObject.params)
            return true
        default:
            throw new Error('Method not found.')
    }
})

async function getChainId() {
    const chainId = await wallet.send({
        method: 'net_version',
    })
    if (!chainId) {
        throw new Error('Can not get chain ID.')
    }
    return chainId
}

async function callBackend(endpoint, payload) {
    const url = `${BACKEND_URL}/${endpoint}/`
    console.log(`sending to ${url}: `, payload)
    const response = await fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
    })
    const result = await response.json()
    if (result.error) {
        throw new Error(result.error)
    }
    console.log('received: ', result)
    return result
}

async function addAccount(daoAddress) {
    const chainId = await getChainId()
    const payload = {
        chainId: chainId,
        dao: daoAddress,
    }
    const { agentAddress } = await callBackend('agent', payload)
    await wallet.send({
        method: 'wallet_manageIdentities',
        params: [
            'add',
            {address: agentAddress},
        ],
    })
    accounts[agentAddress] = {dao: daoAddress, actor: null}
    return agentAddress
}

async function setActor(account, actor) {
    const allowedActors = await wallet.send({
        method: 'eth_accounts',
    })
    if (allowedActors.indexOf(actor) === -1) {
        throw new Error('Account not found.')
    }
    accounts[account].actor = actor
}

wallet.registerAccountMessageHandler(async (originString, requestObject) => {
    switch (requestObject.method) {
        case 'eth_signTransaction':
            await sendTransaction(requestObject.params[0])
            break
        default:
            throw Error('Method not supported.')
    }
    return true
})

async function sendTransaction(tx) {
    const account = tx.from
    const {dao, actor} = accounts[account]
    if (!actor) {
        throw new Error('Actor not set for current account.')
    }
    const chainId = await getChainId()
    const payload = {
        chainId: chainId,
        dao: dao,
        actor: actor,
        txParams: [tx.to, tx.value, tx.data],
    }
    const result = await callBackend('path', payload)
    const newTx = result.tx
    newTx.gas = newTx.gas.toString(16)
    console.log('sending transaction: ', newTx)
    const txId = await wallet.send({
        method: 'eth_sendTransaction',
        params: [newTx],
    })
    console.log('transaction ID: ', txId)
}


backend.js
#!/usr/bin/env node

/*
 * Supported environment variables:
 * - ETHEREUM_URL
 * - IPFS_GATEWAY_URL
 */

const express = require('express')
const agent = require('./agent')

const app = express()
app.use(express.json())

app.post('/agent/', async (request, response) => {
    console.log('Incoming request: ', request.body)
    try {
        const {chainId, daoName} = request.body
        const result = await agent.findAgent(chainId, daoName)
        response.json(result)
    } catch (error) {
        console.error(error)
        response.json({error: error.toString()})
    }
})

app.post('/path/', async (request, response) => {
    console.log('Incoming request: ', request.body)
    try {
        const {chainId, dao, actor, txParams} = request.body
        const wrappedTx = await agent.calculatePath(chainId, dao, actor, txParams)
        response.json({ wrappedTx })
    } catch (error) {
        console.error(error)
        response.json({error: error.toString()})
    }
})

const PORT = 8082

app.listen(PORT, () => {
    console.log(`Backend service running on: http://localhost:${PORT}`)
})


plugin.js
const BACKEND_URL = 'http://localhost:8082'

const accounts = {}

wallet.registerRpcMessageHandler(async (originString, requestObject) => {
    switch (requestObject.method) {
        case 'addAccount':
            return await addAccount(requestObject.params[0])
        case 'setActor':
            await setActor(...requestObject.params)
            return true
        default:
            throw new Error('Method not found.')
    }
})

async function getChainId() {
    const chainId = await wallet.send({
        method: 'net_version',
    })
    if (!chainId) {
        throw new Error('Can not get chain ID.')
    }
    return chainId
}

async function callBackend(endpoint, payload) {
    const url = `${BACKEND_URL}/${endpoint}/`
    console.log(`sending to ${url}: `, payload)
    const response = await fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
    })
    const result = await response.json()
    if (result.error) {
        throw new Error(result.error)
    }
    console.log('received: ', result)
    return result
}

async function addAccount(daoName) {
    const chainId = await getChainId()
    const payload = {
        chainId: chainId,
        daoName: daoName,
    }
    const { daoAddress, agentAddress } = await callBackend('agent', payload)
    await wallet.send({
        method: 'wallet_manageIdentities',
        params: [
            'add',
            {address: agentAddress},
        ],
    })
    accounts[agentAddress] = {dao: daoAddress, actor: null}
    return agentAddress
}

async function setActor(account, actor) {
    const allowedActors = await wallet.send({
        method: 'eth_accounts',
    })
    if (allowedActors.indexOf(actor) === -1) {
        throw new Error('Account not found.')
    }
    accounts[account].actor = actor
}

wallet.registerAccountMessageHandler(async (originString, requestObject) => {
    switch (requestObject.method) {
        case 'eth_signTransaction':
            return await sendTransaction(requestObject.params[0])
        default:
            throw Error('Method not supported.')
    }
})

async function sendTransaction(tx) {
    const account = tx.from
    const {dao, actor} = accounts[account]
    if (!actor) {
        throw new Error('Actor not set for current account.')
    }
    const chainId = await getChainId()
    const payload = {
        chainId: chainId,
        dao: dao,
        actor: actor,
        txParams: [tx.to, tx.value, tx.data],
    }
    const { wrappedTx } = await callBackend('path', payload)
    wrappedTx.gas = wrappedTx.gas.toString(16)
    console.log('sending transaction: ', wrappedTx)
    const wrappedTxId = await wallet.send({
        method: 'eth_sendTransaction',
        params: [wrappedTx],
    })
    console.log('transaction ID: ', wrappedTxId)
    return wrappedTxId
}
